sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/InferFiltersFromConstraintsSuite.scala

```diff
@@ -192,4 +192,17 @@ class InferFiltersFromConstraintsSuite extends PlanTest {
 
     comparePlans(Optimize.execute(original.analyze), correct.analyze)
   }
+
+  test("SPARK-23405:single left-semi join, filter out nulls on either side on equi-join keys") {
+    val x = testRelation.subquery('x)
+    val y = testRelation.subquery('y)
+    val originalQuery = x.join(y, LeftSemi,
+      condition = Some("x.a".attr === "y.a".attr)).analyze
```

Comment:
nit: we can create a `val condition = Some("x.a".attr === "y.a".attr)` to reduce duplicated code